features/dd-sdk-android-trace/src/test/kotlin/com/datadog/trace/core/CoreSpanBuilderTest.kt

```diff
@@ -0,0 +1,534 @@
+
+        assertThat(span.tags).isEqualTo(
+            mapOf(
+                DDTags.THREAD_NAME to Thread.currentThread().name,
+                DDTags.THREAD_ID to Thread.currentThread().id,
+                DDTags.RUNTIME_ID_TAG to Config.get().runtimeId,
+                DDTags.LANGUAGE_TAG_KEY to DDTags.LANGUAGE_TAG_VALUE,
+                DDTags.PID_TAG to Config.get().processId,
+                DDTags.SCHEMA_VERSION_TAG_KEY to SpanNaming.instance().version(),
+                DDTags.PROFILING_ENABLED to if (Config.get().isProfilingEnabled()) 1 else 0
+            )
+        )
+
+        // with all custom fields provided
+        val expectedResource = "fakeResource"
+        val expectedService = "fakeService"
+        val expectedType = "fakeType"
+
+        span = tracer
+            .buildSpan(instrumentationName, expectedName)
+            .withServiceName("foo")
+            .withResourceName(expectedResource)
+            .withServiceName(expectedService)
+            .withErrorFlag()
+            .withSpanType(expectedType)
+            .start()
+
+        val context = span.context() as DDSpanContext
+
+        assertThat(context.resourceName).isEqualTo(expectedResource)
+        assertThat(context.errorFlag).isTrue
+        assertThat(context.serviceName).isEqualTo(expectedService)
+        assertThat(context.spanType).isEqualTo(expectedType)
+
+        assertThat(context.getTag(DDTags.THREAD_NAME)).isEqualTo(Thread.currentThread().name)
+        assertThat(context.getTag(DDTags.THREAD_ID)).isEqualTo(Thread.currentThread().id)
+    }
+
+    @Test
+    fun `setting name should remove`() {
+        val nameValues = listOf(
+            Pair("null.tag", null),
+            Pair("empty.tag", "")
+        )
+
+        for ((name, value) in nameValues) {
+            val span = tracer.buildSpan(instrumentationName, "op name")
+                .withTag(name, "tag value")
+                .withTag(name, value)
+                .start()
+
+            assertThat(span.tags[name]).isNull()
+
+            span.setTag(name, "a tag")
+
+            assertThat(span.tags[name]).isEqualTo("a tag")
+
+            span.setTag(name, value)
+
+            assertThat(span.tags[name]).isNull()
+        }
+    }
+
+    @Test
+    fun `should build span timestamp in nano`() {
+        // time in micro
+        val expectedTimestamp = 487517802L * 1000 * 1000L
+        val expectedName = "fakeName"
+
+        var span = tracer
+            .buildSpan(instrumentationName, expectedName)
+            .withServiceName("foo")
+            .withStartTimestamp(expectedTimestamp)
+            .start()
+
+        // get return nano time
+        assertThat(span.startTime).isEqualTo(expectedTimestamp * 1000L)
+
+        // auto-timestamp in nanoseconds
+        val start = System.currentTimeMillis()
+        span = tracer.buildSpan(instrumentationName, expectedName).withServiceName("foo").start()
```

Comment:
```suggestion
        val secondSpan = tracer.buildSpan(instrumentationName, expectedName).withServiceName("foo").start()
```

same suggestion applies to other tests - instead of using `var` and changing the reference, it is better to be more explicit in saying that it is another instance.